### PR TITLE
PS-8173 - Verify that there are no TokuDB tables before upgrade on server start

### PR DESCRIPTION
**Pull request by ldonoso**
Check that here are no TokuDB tables in the schema before allowing a
potentially destructive upgrade to occur.

The check was already implemented for upgrades from 5.7.

### CODE REVIEW

File: sql/dd/impl/upgrade/server.cc [hunk 1]
```diff
@@ -560,6 +560,15 @@ static bool check_tables(THD *thd, std::unique_ptr<Schema> &schema,
   auto process_table = [&](std::unique_ptr<dd::Table> &table) {
     invalid_triggers(thd, schema->name().c_str(), *table);
 
+    // The TokuDB engine was removed in 8.0.28 Don't upgrade if it is used.
+    if (my_strcasecmp(system_charset_info, table->engine().c_str(), "TokuDB") ==
```
Comment (satya-bodapati) [inline sql/dd/impl/upgrade/server.cc:564]: can we use db_type check with DB_TYPE_TOKUDB?

File: share/messages_to_error_log.txt [hunk 2]
```diff
@@ -11708,6 +11708,9 @@ ER_KEYRING_COMPONENT_NO_CONFIG
 ER_KEYRING_COMPONENT_CONFIG_PARSE_FAILED
   eng "Keyring configuration JSON parse error: %s (%lu)"
 
+ER_PERCONA_UPDATE_NOT_SUPPORTED_ENGINE
+  eng "Removed storage engine '%s' is used in '%s.%s'"
```
Comment (LenzGr) [inline share/messages_to_error_log.txt]: Small textual improvement suggestion (this probably requires re-ordering the parameters passed here and the corresponding tests):
```suggestion
  eng "Table '%s.%s' requires the '%s' storage engine, which has been removed."
```

File: share/messages_to_error_log.txt [hunk 3]
```diff
@@ -11708,6 +11708,9 @@ ER_KEYRING_COMPONENT_NO_CONFIG
 ER_KEYRING_COMPONENT_CONFIG_PARSE_FAILED
   eng "Keyring configuration JSON parse error: %s (%lu)"
 
+ER_PERCONA_UPDATE_NOT_SUPPORTED_ENGINE
```
Comment (satya-bodapati) [inline share/messages_to_error_log.txt]: ER_PERCONA_UNSUPPORTED_ENGINE or just ER_UNSUPPORTED_ENGINE

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: share/messages_to_error_log.txt (hunk 2 above)
Comment (LenzGr) [reply share/messages_to_error_log.txt]: As an example: "'Table 'foo.bar' requires the 'TokuDB' storage engine, which has been removed."

File: sql/dd/impl/upgrade/server.cc (hunk 1 above)
Comment (ldonoso) [reply sql/dd/impl/upgrade/server.cc:564]: @satya-bodapati , I'd prefer to be consistent with the way the engine is checked in the rest of the file. Please, mind how ,some lines below, the check for the engine "InnoDB" is made

File: mysql-test/t/mysql_upgrade_fails_toku_80.test [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+--file_exists $MYSQL_TMP_DIR/$ZIP_BASENAME
+--exec unzip -qo $MYSQL_TMP_DIR/$ZIP_BASENAME -d $MYSQL_TMP_DIR
+--remove_file $MYSQL_TMP_DIR/$ZIP_BASENAME
+
+# Test the upgrade fails when previous is 8.0.x and uses TokuDb
+
+--let $MYSQLD_LOG= $MYSQLTEST_VARDIR/log/mysql_ps8173_upgrade_complete.log
+--replace_result $MYSQLD MYSQLD $MYSQLD_DATADIR1 MYSQLD_DATADIR1 $MYSQLD_LOG MYSQLD_LOG
+--error 1
+--exec $MYSQLD --no-defaults --secure-file-priv="" --datadir=$MYSQLD_DATADIR1 --log-error=$MYSQLD_LOG
```
Comment (satya-bodapati) [inline mysql-test/t/mysql_upgrade_fails_toku_80.test:27]: @ldonoso is the datadir from failed upgrade (to 8.0) usable by 5.7 server? (please do manual check)

File: mysql-test/t/mysql_upgrade_fails_toku_57.test [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+# Stop the default mtr server
+
+--let $shutdown_server_timeout = 300
+--source include/shutdown_mysqld.inc
+
+# Setup
+
+--let $MYSQLD_DATADIR1= $MYSQL_TMP_DIR/data_ps8173_57
+--let $ZIP_BASENAME= data_ps8173_57.zip
+--let $ZIP_FILE= $MYSQLTEST_VARDIR/std_data/$ZIP_BASENAME
```
Comment (satya-bodapati) [inline mysql-test/t/mysql_upgrade_fails_toku_57.test]: if possible, please place them in std_data/upgrade directory

File: mysql-test/t/mysql_upgrade_fails_toku_57.test [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+--copy_file $ZIP_FILE $MYSQL_TMP_DIR/$ZIP_BASENAME
+--file_exists $MYSQL_TMP_DIR/$ZIP_BASENAME
+--exec unzip -qo $MYSQL_TMP_DIR/$ZIP_BASENAME -d $MYSQL_TMP_DIR
+--remove_file $MYSQL_TMP_DIR/$ZIP_BASENAME
+
+# Test the upgrade fails when previous is 5.7.x and uses TokuDb
+
+--let $MYSQLD_LOG= $MYSQLTEST_VARDIR/log/mysql_ps8173_upgrade_complete.log
```
Comment (satya-bodapati) [inline mysql-test/t/mysql_upgrade_fails_toku_57.test:24]: remove this file

File: mysql-test/t/mysql_upgrade_fails_toku_80.test [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+--let $ZIP_FILE= $MYSQLTEST_VARDIR/std_data/$ZIP_BASENAME
+
+--copy_file $ZIP_FILE $MYSQL_TMP_DIR/$ZIP_BASENAME
+--file_exists $MYSQL_TMP_DIR/$ZIP_BASENAME
+--exec unzip -qo $MYSQL_TMP_DIR/$ZIP_BASENAME -d $MYSQL_TMP_DIR
+--remove_file $MYSQL_TMP_DIR/$ZIP_BASENAME
+
+# Test the upgrade fails when previous is 8.0.x and uses TokuDb
+
+--let $MYSQLD_LOG= $MYSQLTEST_VARDIR/log/mysql_ps8173_upgrade_complete.log
```
Comment (satya-bodapati) [inline mysql-test/t/mysql_upgrade_fails_toku_80.test:24]: remove this file at end of test

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: sql/dd/impl/upgrade/server.cc (hunk 1 above)
Comment (satya-bodapati) [reply sql/dd/impl/upgrade/server.cc:564]: Fine with me.

**Comment by satya-bodapati** [comment]
@ldonoso  please fix bug title in github & commit message.

and submit these two test on Jenkins single platform (pipeline)

File: mysql-test/t/mysql_upgrade_fails_toku_80.test (hunk 4 above)
Comment (ldonoso) [reply mysql-test/t/mysql_upgrade_fails_toku_80.test:27]: @satya-bodapati , it works correctly. To find out I have:

- Modified the mtr test so we do not delete the data dir after running `mysql_upgrade_fails_toku_80.test`
- Initiated an instance of 5.7 using the data dir generated in the test above
- Connected to the instance and review that the schemas are working

**Comment by satya-bodapati** [comment]
https://ps80.cd.percona.com/view/8.0/job/percona-server-8.0-pipeline/38121/

**Review by satya-bodapati — APPROVED** [review]

**Review by george-lorch — COMMENTED** [review]
I'm a little late, but LGTM after the suggested changes.  Thank you!

**Pull request merged**